lib/tasks/worker.rake

```diff
@@ -1,13 +1,108 @@
 # frozen_string_literal: true
 
+require("concurrent")
+
+MAX_DELAY = 1.minutes
+
 namespace(:worker) do
-  desc("Run the worker")
+  # if timestamp already passed or same as Time.now we will process the message
+  def to_process?(timestamp)
+    remaining_time = [timestamp.to_i - Time.now.to_i, 0].max
+    remaining_time.zero?
+  end
+
+  # execution of job and acknowledging of message after execution
+  def process_message(message, pubsub)
+    puts("\n [WORKER][process_message] Time to process.")
+    message.modify_ack_deadline!(MAX_DELAY.to_i)
+
+    ActiveJob::Base.execute(JSON.parse(message.data))
+
+    # puts("\n [WORKER][process_message] About to acknowledge message #{message.message_id}.")
+    message.acknowledge!
+    puts("\n [WORKER][process_message] Message #{message.message_id} was acknowledged!")
+  rescue StandardError => e
+    puts("\n [ERROR][WORKER][process_message]: #{e.message} #{e.backtrace}")
+    enqueue_failed_jobs(message, pubsub)
+  end
+
+  # publish failed jobs to morgue queue
+  # create a subscription under the failed jobs topic
+  def enqueue_failed_jobs(message, pubsub)
+    puts("\n [WORKER][enqueue_failed_jobs] Message: #{message.data}")
+
+    failed_topic = "#{topic_name}-morgue-queue"
+    pubsub.topic(failed_topic).publish(JSON.dump(message.data)) # publish failed messages to morgue topic
+    pubsub.subscribe(failed_topic) # create a subscription
+    puts("\n [WORKER][enqueue_failed_jobs] Done publishing: #{message.data}")
+  rescue StandardError => e
+    puts("\n [ERROR][WORKER][enqueue_failed_jobs]: Error publishing to morgue queue : #{e.message}")
+  end
+
+  # use the environment variable set for the topic name
+  # to make it configurable
+  def topic_name
+    @topic_name ||= Rails.application.topic_name
+  end
+
+  desc('Run the worker')
   task(run: :environment) do
     # See https://googleapis.dev/ruby/google-cloud-pubsub/latest/index.html
 
     puts("Worker starting...")
 
-    # Block, letting processing threads continue in the background
+    pubsub = Pubsub.new
+    subscription = pubsub.subscribe(topic_name)
+
+    subscriber = subscription.listen do |message|
+      puts("\n Received message: #{message.data}")
+      if to_process?(message.attributes['timestamp'])
+        puts("\n [LISTEN] Processing right away....")
+        process_message(message, pubsub)
+      else
+        delay = message.attributes['timestamp'].to_i - Time.now.to_i
```

Comment:
What is the max delay?

Reply:
There is no max delay if it enters the else condition, I added this commit https://github.com/skysemilla/kisi-api-challenge/pull/1/commits/ec6213b392d2a84d95cba44d039a2f74c5d66d7f to use the MAX_DELAY to provide a cap when delaying the messages.